{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spreadsheet validation with pandas\n",
    "\n",
    "Once upon a time I was asked to manually, as in just by eyeballing it, validate some bioinformatics data in a spreadsheet to spot rows that did not conform to a given set of rules. Yes, that's a silly thing to ask someone to do. It's a long story, don't ask. \n",
    "\n",
    "Anyway, I'd been looking for opportunities to play around with data processing in [pandas](https://pandas.pydata.org/), and decided that taking this request to the next level would be a great way to do that. It's just a first pass, and I'll discuss things I would have done differently at the end, but I'm happy enough with the results to put them up here. \n",
    "\n",
    "Enjoy!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First steps\n",
    "\n",
    "Let's take care of some imports, load up the spreadsheet, and see what we're working with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "    Row Primary SGDID Feature type Feature name Gene name Secondary SGDID  \\\n",
       "0     1    S000001568          ORF      YKL085W      MDH1      L000001045   \n",
       "1     2    S000001569          ORF      YKL086W      SRX1                   \n",
       "2     7    S000001574          ORF      YKL091C                             \n",
       "3     8    S000001575          ORF      YKL092C      BUD2      L000000199   \n",
       "4     9    S000001576          ORF      YKL093W      MBR1      L000001033   \n",
       "..  ...           ...          ...          ...       ...             ...   \n",
       "79  147    S000001705          ORF      YKL222C                             \n",
       "80  148    S000001706          ORF      YKL223W                             \n",
       "81  149    S000001707          ORF      YKL224C     PAU16                   \n",
       "82  150    S000001708          ORF      YKL225W                             \n",
       "83  151    S000001729          ORF      YKR021W      ALY1                   \n",
       "\n",
       "   Chromosome  Start coordinate  Stop coordinate Strand Sequence version date  \\\n",
       "0          11            279123           280127      W            1996-07-31   \n",
       "1          11            278281           278664      W            1996-07-31   \n",
       "2          11            270650           269718      C            1996-07-31   \n",
       "3          11            269459           266145      C            1996-07-31   \n",
       "4          11            264789           265808      W            1996-07-31   \n",
       "..        ...               ...              ...    ...                   ...   \n",
       "79         11              5620             3503      C            1996-07-31   \n",
       "80         11              2389             2721      W            1996-07-31   \n",
       "81         11              2181             1810      C            1996-07-31   \n",
       "82         11               451              798      W            1996-07-31   \n",
       "83         11            479234           481981      W            1996-07-31   \n",
       "\n",
       "                                          Description Unnamed: 12  \n",
       "0   Mitochondrial malate dehydrogenase; catalyzes ...              \n",
       "1   Sulfiredoxin; contributes to oxidative stress ...              \n",
       "2   Putative phosphatidylinositol/phosphatidylchol...              \n",
       "3   GTPase activating factor for Rsr1p/Bud1p; play...              \n",
       "4   protein involved in mitochondrial functions an...              \n",
       "..                                                ...         ...  \n",
       "79  Protein of unknown function; may interact with...              \n",
       "80  Dubious open reading frame; unlikely to encode...              \n",
       "81  Protein of unknown function; member of the ser...              \n",
       "82  Dubious open reading frame; unlikely to encode...              \n",
       "83  Alpha arrestin, substrate of calcineurin; cont...              \n",
       "\n",
       "[84 rows x 13 columns]"
      ],
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Row</th>\n      <th>Primary SGDID</th>\n      <th>Feature type</th>\n      <th>Feature name</th>\n      <th>Gene name</th>\n      <th>Secondary SGDID</th>\n      <th>Chromosome</th>\n      <th>Start coordinate</th>\n      <th>Stop coordinate</th>\n      <th>Strand</th>\n      <th>Sequence version date</th>\n      <th>Description</th>\n      <th>Unnamed: 12</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1</td>\n      <td>S000001568</td>\n      <td>ORF</td>\n      <td>YKL085W</td>\n      <td>MDH1</td>\n      <td>L000001045</td>\n      <td>11</td>\n      <td>279123</td>\n      <td>280127</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>Mitochondrial malate dehydrogenase; catalyzes ...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>2</td>\n      <td>S000001569</td>\n      <td>ORF</td>\n      <td>YKL086W</td>\n      <td>SRX1</td>\n      <td></td>\n      <td>11</td>\n      <td>278281</td>\n      <td>278664</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>Sulfiredoxin; contributes to oxidative stress ...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>7</td>\n      <td>S000001574</td>\n      <td>ORF</td>\n      <td>YKL091C</td>\n      <td></td>\n      <td></td>\n      <td>11</td>\n      <td>270650</td>\n      <td>269718</td>\n      <td>C</td>\n      <td>1996-07-31</td>\n      <td>Putative phosphatidylinositol/phosphatidylchol...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>8</td>\n      <td>S000001575</td>\n      <td>ORF</td>\n      <td>YKL092C</td>\n      <td>BUD2</td>\n      <td>L000000199</td>\n      <td>11</td>\n      <td>269459</td>\n      <td>266145</td>\n      <td>C</td>\n      <td>1996-07-31</td>\n      <td>GTPase activating factor for Rsr1p/Bud1p; play...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>9</td>\n      <td>S000001576</td>\n      <td>ORF</td>\n      <td>YKL093W</td>\n      <td>MBR1</td>\n      <td>L000001033</td>\n      <td>11</td>\n      <td>264789</td>\n      <td>265808</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>protein involved in mitochondrial functions an...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>79</th>\n      <td>147</td>\n      <td>S000001705</td>\n      <td>ORF</td>\n      <td>YKL222C</td>\n      <td></td>\n      <td></td>\n      <td>11</td>\n      <td>5620</td>\n      <td>3503</td>\n      <td>C</td>\n      <td>1996-07-31</td>\n      <td>Protein of unknown function; may interact with...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>80</th>\n      <td>148</td>\n      <td>S000001706</td>\n      <td>ORF</td>\n      <td>YKL223W</td>\n      <td></td>\n      <td></td>\n      <td>11</td>\n      <td>2389</td>\n      <td>2721</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>Dubious open reading frame; unlikely to encode...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>81</th>\n      <td>149</td>\n      <td>S000001707</td>\n      <td>ORF</td>\n      <td>YKL224C</td>\n      <td>PAU16</td>\n      <td></td>\n      <td>11</td>\n      <td>2181</td>\n      <td>1810</td>\n      <td>C</td>\n      <td>1996-07-31</td>\n      <td>Protein of unknown function; member of the ser...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>82</th>\n      <td>150</td>\n      <td>S000001708</td>\n      <td>ORF</td>\n      <td>YKL225W</td>\n      <td></td>\n      <td></td>\n      <td>11</td>\n      <td>451</td>\n      <td>798</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>Dubious open reading frame; unlikely to encode...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>83</th>\n      <td>151</td>\n      <td>S000001729</td>\n      <td>ORF</td>\n      <td>YKR021W</td>\n      <td>ALY1</td>\n      <td></td>\n      <td>11</td>\n      <td>479234</td>\n      <td>481981</td>\n      <td>W</td>\n      <td>1996-07-31</td>\n      <td>Alpha arrestin, substrate of calcineurin; cont...</td>\n      <td></td>\n    </tr>\n  </tbody>\n</table>\n<p>84 rows Ã— 13 columns</p>\n</div>"
     },
     "metadata": {},
     "execution_count": 12
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import datetime, re\n",
    "import np\n",
    "from collections import defaultdict\n",
    "\n",
    "data = pd.ExcelFile(\"SGD-Features.xlsx\")\n",
    "sheet = data.parse().fillna('')\n",
    "sheet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We need to apply the following validation rules:\n",
    "```\n",
    "1.      Row number: use this when you refer to the rows with errors (mandatory)\n",
    "2.   \tPrimary SGDID: a unique identifier for the sequence feature (mandatory)\n",
    "3.   \tFeature type: a phrase that describes the sequence feature   (mandatory\n",
    "4.   \tFeature name: a name that describes the approximate location of the sequence feature (optional)\n",
    "5.      Gene name: a human-friendly name, following the pattern of 3-4 letters and a number (optional)\n",
    "6.   \tSecondary SGDID: alternate identifiers for the sequence feature  (optional, multiples separated by |)\n",
    "7.   \tChromosome: the chromosome that the sequence feature is located on (optional)\n",
    "8.  \tStart coordinate: the coordinate number on the chromosome that contains the beginning of the start codon (mandatory)\n",
    "9.  \tStop coordinate: the coordinate number on the chromosome that  contains the end of the stop codon  (mandatory)\n",
    "10.  \tStrand: whether the ORF is on the Watson or Crick strand (mandatory)\n",
    "11.  \tSequence version date: date when the genomic sequence was last modified (optional)\n",
    "12.  \tDescription: A short summary of the biological role of the ORF (optional)\n",
    "```\n",
    "\n",
    "## Or, put another way...\n",
    "\n",
    "| Field name                 | Is required? | Content must be...\n",
    "|---                         |---           |---\n",
    "| Row                        | Y              | a positive integer  \n",
    "| Primary SGDID              | Y            | unique, an SGID\n",
    "| Feature type               | Y            |  \n",
    "| Feature name               | N            | \n",
    "| Gene name                  | N            | 3-4 letters followed by a number\n",
    "| Secondary SGDID            | N            | '\\|'-delimited SGIDs\n",
    "| Chromosome                 | N            | a positive integer\n",
    "| Start coordinate           | Y            | a positive integer\n",
    "| Stop coordinate            | Y            | a positive integer\n",
    "| Strand                     | Y            | \"W\" or \"C\"\n",
    "| Sequence version date      | N            | date\n",
    "| Description                | N            | \n",
    "\n",
    "## Let's write a function for each of those validations\n",
    "\n",
    "Each one takes a pandas dataframe and a column name, and returns a dataframe with True/False values for each index in the original, corresponding to whether the value for the given column at that index passed validaion. \n",
    "This is handy because then multiple validator results can be combined with bitwise boolean operations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell values we will count as \"empty\"\n",
    "EMPTY_VALS = [\"\",np.nan,None]\n",
    "\n",
    "def exists(df, col, required = True):\n",
    "    \"must not be blank\"\n",
    "    # In this case the `required` option is really just \n",
    "    # here for consistency with other validators, \n",
    "    # to make testing easier\n",
    "    if required:\n",
    "        return ~df[col].isin(EMPTY_VALS)\n",
    "    else:\n",
    "        return df[col].apply(lambda x: True)\n",
    "\n",
    "\n",
    "# helper function controlling whether the calling \n",
    "# validator tolerates empty cells\n",
    "def _with_optionality(df, col, matches, required):\n",
    "    if required:\n",
    "        return matches & exists(df, col)\n",
    "    else:\n",
    "        return matches | ~exists(df, col)\n",
    "    \n",
    "    \n",
    "def is_positive_int(df, col, required = True):\n",
    "    \"must be a positive integer\"\n",
    "    # corece everything to a number or NaN\n",
    "    as_nums = pd.to_numeric(df[col], errors='coerce')\n",
    "    \n",
    "    # identify positive \"numbers\" (includes converted strings like \"1\")\n",
    "    are_positive = as_nums > 0\n",
    "    are_ints = as_nums % 1 == 0\n",
    "    \n",
    "    # for filtering out non-integers, including strings as described above\n",
    "    are_actually_numbers = df[col].apply(lambda x: pd.api.types.is_number(x))\n",
    "    \n",
    "    matches = are_positive & are_ints & are_actually_numbers\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "    \n",
    "    \n",
    "def is_unique(df, col, required=True):\n",
    "    \"must be unique\"\n",
    "    # idx,bool: True only for appearance 2+ of a val, not originals\n",
    "    duplicate_idx = df[col].duplicated()\n",
    "\n",
    "    # rows with appearance 2+. still doesn't include the original.\n",
    "    # to get that, we'll need the entire row so we can get...\n",
    "    duplicate_rows = df[duplicate_idx]\n",
    "\n",
    "    # SGDID of ^^^\n",
    "    duplicate_sgdids = duplicate_rows[col]\n",
    "\n",
    "    # idx, bool: True for all rows containing a duplicated SGDID\n",
    "    all_duplicates_idx = df[col].isin(duplicate_sgdids)\n",
    "\n",
    "    matches = ~all_duplicates_idx\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "\n",
    "\n",
    "def _matches_regex(df, col, regex):\n",
    "    return df[col].apply(lambda v: bool(re.match(regex, str(v))))\n",
    "\n",
    "\n",
    "def is_gene_friendly_name(df, col, required=False):\n",
    "    \"must be 3-4 letters followed by a number\"\n",
    "    matches = _matches_regex(df, col, r'^[A-Z]{3,4}[0-9]+$')\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "\n",
    "\n",
    "def is_valid_sgdid(df, col, required=True):\n",
    "    \"must be L or S followed by 9 numbers\"\n",
    "    matches = _matches_regex(df, col, r'^[LS][0-9]{9}$')\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "\n",
    "\n",
    "def is_valid_sgdid_list(df, col, required=False):\n",
    "    \"must be a '|'-delimited list of SGDIDs\"\n",
    "    # table of is_valid_sgdid results for each '|'-delimited val, row for row\n",
    "    matches = df[col].apply(lambda v: is_valid_sgdid(pd.DataFrame({col: str(v).split('|')}), col))\n",
    "    \n",
    "    # rows with no False validity booleans\n",
    "    matches = matches.fillna(True).all(axis=1)\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "\n",
    "\n",
    "def is_valid_strand(df, col, required=True):\n",
    "    \"must be 'W' or 'C'\"\n",
    "    matches = df[col].isin(['C','W'])\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "\n",
    "\n",
    "def is_valid_date(df, col, required=True):\n",
    "    \"must be a valid date in YYYY-MM-DD format\"\n",
    "    def check_date(date_str):\n",
    "        try:\n",
    "            datetime.datetime.strptime(str(date_str), '%Y-%m-%d')\n",
    "        except ValueError:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "        \n",
    "    matches = df[col].apply(lambda v: check_date(v))\n",
    "    return _with_optionality(df, col, matches, required)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests for the validators\n",
    "\n",
    "Before applying validators to the spreadsheet, let's include some tests to make sure they work as expected (plus the tests themselves can provide readers with more insight into each validator's intended function)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "exists (required)... PASS\nexists (optional)... PASS\nis_positive_int (required)... PASS\nis_positive_int (optional)... PASS\nis_unique (required)... PASS\nis_unique (optional)... PASS\nis_gene_friendly_name (required)... PASS\nis_gene_friendly_name (optional)... PASS\nis_valid_sgdid (required)... PASS\nis_valid_sgdid (optional)... PASS\nis_valid_sgdid_list (required)... PASS\nis_valid_sgdid_list (optional)... PASS\nis_valid_strand (required)... PASS\nis_valid_strand (optional)... PASS\nis_valid_date (required)... PASS\nis_valid_date (optional)... PASS\n"
     ]
    }
   ],
   "source": [
    "validator_tests = {\n",
    "    exists: (\n",
    "        [1,    \"a\",  \" \" ],  # input\n",
    "        [True, True, True]   # expected output\n",
    "    ),\n",
    "    is_positive_int: (\n",
    "        [1,    \"1\",   1.5,   -1,  ],\n",
    "        [True, False, False, False]\n",
    "    ),\n",
    "    is_unique: (\n",
    "        [1,     2,    'a',   'b',  1,     'a'   ],\n",
    "        [False, True, False, True, False, False ]\n",
    "    ),\n",
    "    is_gene_friendly_name: (\n",
    "        [\"ABC1\", \"ABCD2\", \"ABCD22\", \"ABC2a\", \"ABC\",  'AB1',  '1',   'abc1'],\n",
    "        [True,   True,    True,     False,   False,  False,  False, False ]\n",
    "    ),\n",
    "    is_valid_sgdid: (\n",
    "        [\"L123456789\", \"S123456789\", \"A123456789\", \"L12345678\", \"L1234567890\"],\n",
    "        [True,         True,         False,        False,       False         ]\n",
    "    ),\n",
    "    is_valid_sgdid_list: (\n",
    "        [\"L123456789\", \"S123456789|L123456789\", \"S123456789|A123456789\", \"S123456789,L123456789\", ],\n",
    "        [True,         True,                    False,                   False,                   ]\n",
    "    ),\n",
    "    is_valid_strand: (\n",
    "        ['C',   'W',  'c',   'X',  ],\n",
    "        [True, True, False, False ]\n",
    "    ),\n",
    "    is_valid_date: (\n",
    "        ['2020-04-01', '2020-90-01',  '2020-04-90',  '04-09-2020',  ],\n",
    "        [True,         False,         False,         False,         ]\n",
    "    ),\n",
    "}\n",
    "for f, (given, expected) in validator_tests.items():\n",
    "    def run_test(f, given, expected, required):\n",
    "        df = pd.DataFrame({'vals': given})\n",
    "        res = list(f(df, 'vals', required = required))\n",
    "        label = \"{} ({})\".format(f.__name__, \"required\" if required else \"optional\")\n",
    "        if res == expected:\n",
    "            print(\"{}... PASS\".format(label))\n",
    "        else:\n",
    "            report = pd.DataFrame({'Given': given, 'Expected': expected, 'Got': res})\n",
    "            print(\"{}... FAIL\\n    {}\".format(\n",
    "                label,\n",
    "                '\\n    '.join(report.to_string(index=False).split('\\n'))\n",
    "            ))\n",
    "        \n",
    "    given += EMPTY_VALS\n",
    "    expected_when_required = expected + [False] * len(EMPTY_VALS)\n",
    "    expected_when_optional = expected + [True]  * len(EMPTY_VALS)\n",
    "    \n",
    "    run_test(f, given, expected_when_required, True)\n",
    "    run_test(f, given, expected_when_optional, False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now we're ready to associate tests with columms\n",
    "...plus a function to run them. The magic happens in that `for test in tests` loop where, because pandas lets us run the test on an entire column at once, a simple bitwise NOT (`~`) on the results lets us easily separate out the failures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_tests(tests_by_col):\n",
    "    # See below for example tests_by_col data\n",
    "    # There's probably a way to do this in pandas with MultiIndex?\n",
    "    fails_by_row_by_col = defaultdict(lambda: defaultdict(list))\n",
    "\n",
    "    for col, (required, tests) in tests_by_col.items():\n",
    "        col_key = \"{} field {}\".format(\"required\" if required else \"optional\", col)\n",
    "        for test in tests:\n",
    "            valid_idxs = test(sheet, col, required = required)\n",
    "            fail_idxs  = ~valid_idxs\n",
    "            fail_items = sheet[fail_idxs]\n",
    "            if not fail_items.empty:\n",
    "                for row_key in list(fail_items[['Row',col]].apply(lambda df: \"Row {} with value '{}'\".format(df['Row'], df[col]), axis=1)):\n",
    "                    fails_by_row_by_col[col_key][row_key].append(test.__doc__)\n",
    "            \n",
    "    for col_key, fails_by_row in fails_by_row_by_col.items():\n",
    "        print(\"In {}...\".format(col_key))\n",
    "        for row_key, failed_tests in fails_by_row.items():\n",
    "            print (\"  {} failed validation:\\n    - {}\".format(row_key, '\\n    - '.join(failed_tests)))\n",
    "        print('')\n",
    "\n",
    "# Now we associate each column in the spreadsheet with whether it's required, and the relevant validations\n",
    "tests_by_col = {\n",
    "    'Row': (\n",
    "        True,                 # required\n",
    "        [is_positive_int,],   # tests\n",
    "    ),\n",
    "    'Primary SGDID': (\n",
    "        True,\n",
    "        [is_unique, is_valid_sgdid],\n",
    "    ),\n",
    "    \n",
    "    'Feature type': (\n",
    "        True,\n",
    "        [exists],\n",
    "    ),\n",
    "    \n",
    "    'Gene name': (\n",
    "        False,\n",
    "        [is_gene_friendly_name],\n",
    "    ),\n",
    "    \n",
    "    'Secondary SGDID': (\n",
    "        False,\n",
    "        [is_valid_sgdid_list],\n",
    "    ),\n",
    "    'Chromosome': (\n",
    "        False,\n",
    "        [is_positive_int],\n",
    "    ),\n",
    "    'Start coordinate': (\n",
    "        True,\n",
    "        [is_positive_int],\n",
    "    ),\n",
    "    'Stop coordinate': (\n",
    "        True,\n",
    "        [is_positive_int],\n",
    "    ),\n",
    "    'Strand': (\n",
    "        True,\n",
    "        [is_valid_strand],\n",
    "    ),\n",
    "    'Sequence version date': (\n",
    "        False,\n",
    "        [is_valid_date],\n",
    "    ),\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## And the results!\n",
    "_See \"Final thoughts\" below for why the lone blank SGDID on row 114 fails the uniqueness test._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "In required field Primary SGDID...\n  Row 96 with value 'S000001654' failed validation:\n    - must be unique\n  Row 97 with value 'S000001654' failed validation:\n    - must be unique\n  Row 114 with value '' failed validation:\n    - must be unique\n    - must be L or S followed by 9 numbers\n\nIn optional field Gene name...\n  Row 17 with value 'YKL099C' failed validation:\n    - must be 3-4 letters followed by a number\n  Row 55 with value '2013-10-01 00:00:00' failed validation:\n    - must be 3-4 letters followed by a number\n  Row 80 with value 'RPS27A' failed validation:\n    - must be 3-4 letters followed by a number\n  Row 81 with value 'YKL157W' failed validation:\n    - must be 3-4 letters followed by a number\n\nIn optional field Secondary SGDID...\n  Row 17 with value 'chromosome 11' failed validation:\n    - must be a '|'-delimited list of SGDIDs\n\nIn required field Strand...\n  Row 17 with value '255720' failed validation:\n    - must be 'W' or 'C'\n\nIn optional field Sequence version date...\n  Row 31 with value '07-31-1996' failed validation:\n    - must be a valid date in YYYY-MM-DD format\n\n"
     ]
    }
   ],
   "source": [
    "run_tests(tests_by_col)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final thoughts\n",
    "\n",
    "In retrospect I think adding a `required` option to each validator was an un-necessary complication. If I were to do this again, I would make it so if a column is marked as required then the `exists` validator is run on it, and if that fails then it's reported and we don't bother with other validators. Using required-ness to control each validator's tolerance of empty cells is what caused the one blank SGDID on line 114 to be reported as not unique above. Not strictly accurate, but it highlights the error well enough for a first attempt."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "Python 3.8.3 32-bit",
   "display_name": "Python 3.8.3 32-bit",
   "metadata": {
    "interpreter": {
     "hash": "a88e9126dfcdfad8a6084c785d7583984d1d39cf94cbda40b5a47aab39024caa"
    }
   }
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}